Sources/NIOHTTP1/NIOHTTPClientUpgradeHandler.swift

```diff
@@ -0,0 +1,404 @@
+    ///     any other handler that cannot tolerate receiving non-HTTP data.
+    /// - Parameter upgradeCompletionHandler: A closure that will be fired when HTTP upgrade is complete.
+    public init(upgraders: [NIOHTTPClientProtocolUpgrader],
+                httpHandlers: [RemovableChannelHandler],
+                upgradeCompletionHandler: @escaping (ChannelHandlerContext) -> Void) {
+
+        precondition(upgraders.count > 0, "A minimum of one protocol upgrader must be specified.")
+
+        self.upgraders = upgraders
+        self.httpHandlers = httpHandlers
+        self.upgradeCompletionHandler = upgradeCompletionHandler
+    }
+
+    public func write(context: ChannelHandlerContext, data: NIOAny, promise: EventLoopPromise<Void>?) {
+
+        switch self.upgradeState {
+        
+        case .requestRequired:
+            let updatedData = self.addHeadersToOutboundOut(data: data)
+            context.write(updatedData, promise: promise)
+
+        case .awaitingConfirmationResponse:
+            // Still have full http stack.
+            context.write(data, promise: promise)
+            
+        case .upgraderReady, .upgrading:
+            /// We shouldn't be sending anything else until we get a response.
+            context.fireErrorCaught(NIOHTTPClientUpgradeError.writingToHandlerDuringUpgrade)
+
+        case .upgradingAddingHandlers:
+            // These are most likely messages immediately fired by a new protocol handler.
+            // As that is added last we can just forward them on.
+            context.write(data, promise: promise)
+            
+        case .upgradeComplete:
+            //Upgrade complete and this handler should have been removed from the pipeline.
+            context.fireErrorCaught(NIOHTTPClientUpgradeError.writingToHandlerAfterUpgradeCompleted)
+        case .upgradeFailed:
+            //Upgrade failed and this handler should have been removed from the pipeline.
+            context.fireErrorCaught(NIOHTTPClientUpgradeError.writingToHandlerAfterUpgradeCompleted)
+        }
+    }
+
+    private func addHeadersToOutboundOut(data: NIOAny) -> NIOAny {
+        
+        let interceptedOutgoingRequest = self.unwrapOutboundIn(data)
+        
+        if case .head(var requestHead) = interceptedOutgoingRequest {
+            
+            self.upgradeState = .awaitingConfirmationResponse
+            
+            self.addConnectionHeaders(to: &requestHead)
+            self.addUpgradeHeaders(to: &requestHead)
+            return self.wrapOutboundOut(.head(requestHead))
+        }
+        
+        return data
+    }
+
+    private func addConnectionHeaders(to requestHead: inout HTTPRequestHead) {
+        
+        var connectionValue = "upgrade"
+        
+        for upgrader in self.upgraders where upgrader.requiredUpgradeHeaders.count > 0 {
+            connectionValue.append("," + upgrader.requiredUpgradeHeaders.joined(separator: ","))
```

Comment:
This is a quadratic loop, which it'd be nice to avoid. `connectionValue` should probably be of type `[String]` and this line can use `append(contentsOf:)` rather than joining early.

Admittedly, `self.upgraders` is rarely going to be long, but we should still try to avoid quadratic algorithms where we can.